functions/src/__tests__/index.test.ts

```diff
@@ -0,0 +1,248 @@
+      "Content-Type"
+    );
+  }
+
+  it("fails for invalid http verb", async () => {
+    await rallytoken({ method: "PUT" } as functions.Request<any>, response); // eslint-disable-line @typescript-eslint/no-explicit-any
+    verifyCoreResponseHeaders();
+
+    expect(response.status).toHaveBeenCalledWith(500);
+    expect(send).toHaveBeenCalledWith(
+      "Only POST and OPTIONS methods are allowed."
+    );
+  });
+
+  it("returns allowed headers and methods when called with options verb", async () => {
+    await rallytoken({ method: "OPTIONS" } as functions.Request<any>, response); // eslint-disable-line @typescript-eslint/no-explicit-any
+    verifyCoreResponseHeaders();
+
+    expect(response.set).toHaveBeenCalledWith(
+      "Access-Control-Allow-Methods",
+      "POST"
+    );
+
+    expect(response.set).toHaveBeenCalledWith(
+      "Access-Control-Allow-Headers",
+      "Bearer, Content-Type"
+    );
+
+    expect(response.status).toHaveBeenCalledWith(204);
+    expect(send).toHaveBeenCalledWith("");
+  });
+
+  it("fails when POST is invoked with invalid payload", async () => {
+    await rallytoken({ method: "POST" } as functions.Request<any>, response); // eslint-disable-line @typescript-eslint/no-explicit-any
+    verifyCoreResponseHeaders();
+
+    expect(response.status).toHaveBeenCalledWith(500);
+    expect(send).toHaveBeenCalled();
+  });
+
+  it("handles payload in POST request", async () => {
+    const uid = "fake-uid";
+    const customToken = "fake-custom-token";
+
+    const fakeAuth = {
+      verifyIdToken: jest.fn().mockReturnValue({ uid }),
+      createCustomToken: jest.fn().mockReturnValue(customToken),
+    };
+
+    const fakeAuthRef = jest.fn().mockReturnValue(fakeAuth);
+
+    Object.defineProperty(admin, "auth", {
+      get: function () {
+        return fakeAuthRef;
+      },
+    });
+
+    const idToken = "idToken";
+    const studyId = "study1";
+
+    const validateFn = () => {
+      expect(response.status).toHaveBeenCalledWith(200);
+
+      expect(fakeAuth.verifyIdToken).toHaveBeenCalledWith(idToken);
+      expect(fakeAuth.createCustomToken).toHaveBeenCalledWith(
+        `${studyId}:${uid}`,
+        {
+          firebaseUid: uid,
+          studyId,
+        }
+      );
+
+      expect(send.mock.calls[0][0]).toEqual({ rallyToken: customToken });
+    };
+
+    await rallytoken(
+      {
+        method: "POST",
+        body: JSON.stringify({ idToken, studyId }),
+      } as functions.Request<any>, // eslint-disable-line @typescript-eslint/no-explicit-any
+      response
+    );
+
+    validateFn();
+
+    (response.status as jest.Mock).mockClear();
+    (fakeAuth.verifyIdToken as jest.Mock).mockClear();
+    (fakeAuth.createCustomToken as jest.Mock).mockClear();
+    send.mockClear();
+
+    await rallytoken(
+      {
+        method: "POST",
+        body: { idToken, studyId },
+      } as functions.Request<any>, // eslint-disable-line @typescript-eslint/no-explicit-any
```

Comment:
It'd be nice to declare our types, I'll file an issue, no need to worry about that in this PR.

Reply:
Agreed !